auth/api/v1/api.go

```diff
@@ -257,26 +257,72 @@ func (w Wrapper) DrawUpContract(ctx echo.Context) error {
 }
 
-// CreateJwtBearerToken handles the http request (from from the vendor's EPD/XIS) for creating a JWT bearer token which can be used to retrieve an access token from a remote Nuts node.
-func (w Wrapper) CreateJwtBearerToken(ctx echo.Context) error {
-	requestBody := &CreateJwtBearerTokenRequest{}
+// CreateJwtGrant handles the http request (from from the vendor's EPD/XIS) for creating a JWT bearer token which can be used to retrieve an access token from a remote Nuts node.
+func (w Wrapper) CreateJwtGrant(ctx echo.Context) error {
+	requestBody := &CreateJwtGrantRequest{}
 	if err := ctx.Bind(requestBody); err != nil {
 		return err
 	}
 
-	request := services.CreateJwtBearerTokenRequest{
+	request := services.CreateJwtGrantRequest{
 		Actor:         requestBody.Actor,
 		Custodian:     requestBody.Custodian,
 		IdentityToken: &requestBody.Identity,
 		Service:       requestBody.Service,
 		Subject:       requestBody.Subject,
 	}
-	response, err := w.Auth.OAuthClient().CreateJwtBearerToken(request)
+	response, err := w.Auth.OAuthClient().CreateJwtGrant(request)
 	if err != nil {
 		return core.InvalidInputError(err.Error())
 	}
 
-	return ctx.JSON(http.StatusOK, JwtBearerTokenResponse{BearerToken: response.BearerToken})
+	return ctx.JSON(http.StatusOK, JwtGrantResponse{BearerToken: response.BearerToken})
+}
+
+// RequestAccessToken handles the HTTP request (from the vendor's EPD/XIS) for creating a JWT grant and using it as authorization grant to get an access token from the remote Nuts node.
+func (w Wrapper) RequestAccessToken(ctx echo.Context) error {
+	requestBody := &RequestAccessTokenRequest{}
+	if err := ctx.Bind(requestBody); err != nil {
+		return err
+	}
+
+	request := services.CreateJwtGrantRequest{
+		Actor:         requestBody.Actor,
+		Custodian:     requestBody.Custodian,
+		IdentityToken: &requestBody.Identity,
+		Service:       requestBody.Service,
+		Subject:       requestBody.Subject,
+	}
+
+	jwtGrantResponse, err := w.Auth.OAuthClient().CreateJwtGrant(request)
+	if err != nil {
+		logging.Log().WithError(err).Warn("Failed to create JWT grant")
+		return core.InvalidInputError(err.Error())
+	}
+
+	custodianDID, err := did.ParseDID(requestBody.Custodian)
+	if err != nil {
+		logging.Log().WithError(err).WithField("did", requestBody.Custodian).Warn("Invalid custodian DID")
+		return core.InvalidInputError(err.Error())
+	}
+
+	endpointURL, err := w.Auth.OAuthClient().GetOAuthEndpointURL(requestBody.Service, *custodianDID)
+	if err != nil {
+		logging.Log().WithError(err).Error("Failed to get OAuth Endpoint URL")
+		return fmt.Errorf("unable to find the oauth2 service endpoint of the custodian: %w", err)
+	}
+
+	httpClient := HTTPClient{
+		Timeout: w.Auth.HTTPTimeout(),
+	}
+
+	accessTokenResponse, err := httpClient.CreateAccessToken(endpointURL, jwtGrantResponse.BearerToken)
+	if err != nil {
+		logging.Log().WithError(err).Error("Failed to create access token")
+		return fmt.Errorf("unable to create access token: %w", err)
```

Comment:
> 400/403s should be translated to 400.

Currently, I'm using the `core.TestResponseCode` method to check if the request succeeded. I can add a custom error type which also includes the actual HTTP response status and use that here. Is that what you mean?

Reply:
Yes, there's already a custom error (with status code) in core somewhere.

Reply:
I fixed this but I'm unable to test it in the unittests without exporting one of the helper functions in the core package as I'm unable to parse the HTTP response status code in the OAuth API unittests.

Reply:
I think it would be ok to export `httpStatusCodeError` and then use `errors.Is` in the unit tests.
@reinkrul your opinion?

Reply:
In other APIs we test both the actual error and the resolved status code (when using custom errors);

```
assert.ErrorIs(t, err, services.ErrSessionNotFound)
assert.Equal(t, http.StatusNotFound, ctx.wrapper.ResolveStatusCode(err))
```
(auth/api/v1/api_test.go:171)

I think that pattern could be applied here as well?

Reply:
> I think that pattern could be applied here as well?

I don't think that will work as the `ResolveStatusCode` method uses a hardcoded error type to HTTP statuscode mapping. This won't work in this case as it's the OAuth API HTTP client which returns an error. In the HTTP client the error from the upstream API is passed through in the `RequestToken` endpoint which is why we need to parse the actual HTTP statuscode from the response.

Reply:
Since it's a bit hard to map different statusCodes from a downstream call. What about returning a single status code for all downstream errors and mention the error in the description? ("Authorization server returned X")

I don't know the exact suitable statusCode for this, maybe just a 400 to keep it simple?

Reply:
Also the authorization server should return errors according to rfc6749 §5.2. These should be parsed and included in the return.